### PR TITLE
feat(ui-react-storage): add loadingElement to StorageImage

### PR DESCRIPTION
**Pull request by aadimch**
#### Description of changes
Currently, when using StorageImage, the user would see the fallback text while the image is still being fetched. The issue is more prominent when the network is throttled.

This PR adds an `loadingElement` to `StorageImageProps` and uses the loading state exposed by `useGetUrl` and returns the `loadingElement` when loading, if provided. 

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by aadimch** [comment]
Previous [PR](https://github.com/aws-amplify/amplify-ui/pull/6426)

File: packages/react-storage/src/components/StorageImage/StorageImage.tsx [hunk 1]
```diff
@@ -99,7 +100,11 @@ export const StorageImage = ({
     ]
   );
 
-  const { url } = useGetUrl(input);
+  const { url, isLoading } = useGetUrl(input);
+
+  if (isLoading && loadingElement !== undefined) {
+    return loadingElement;
+  }
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageImage/StorageImage.tsx:107]: Leaning towards returning `null` if `isLoading` is `true` and no `loadingElement` is provided. What do you think @dbanksdesign

Comment (aadimch) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx:107]: I would like to add that if we plan to return null if `isLoading` is `true`, we do similar to what I had in the previous [PR](https://github.com/aws-amplify/amplify-ui/pull/6426). 

```
if (isLoading) {
    return <View style={{ ...style }}></View>;
 }
```

If a user provides an `style` with fixed width and height, returning null will cause an layout shift in their application.

File: docs/src/pages/[platform]/connected-components/storage/storageimage/props.ts [hunk 2]
```diff
@@ -40,6 +40,12 @@ export const STORAGE_IMAGE = [
       'A fallback image source to be loaded when the component fails to load the image from Storage',
     type: 'string',
   },
+  {
+    name: 'loadingElement?',
+    description:
+      'A React element to be loaded when the component is loading the image from Storage.',
+    type: 'string',
```
[the remark this answers is not in the document]
Comment (aadimch) [reply docs/src/pages/[platform]/connected-components/storage/storageimage/props.ts:47]: JSX.Element. The StorageImage returns `React.JSX.Element`, so, when I tried to add `loadingElement` type as `React.ReactNode` typescript wasn't happy 😅

[the remark this answers is not in the document]
Comment (calebpollman) [reply docs/src/pages/[platform]/connected-components/storage/storageimage/props.ts:47]: Should be `React.ReactNode`, maybe you need to update the property on the `StorageImageProps` interface?

[the remark this answers is not in the document]
Comment (aadimch) [reply docs/src/pages/[platform]/connected-components/storage/storageimage/props.ts:47]: I changed the loadingElement type to `React.ReactNode`, would returning like this be safe, and is there a reason `React.ReactNode` is preferred over `JSX.Element`?

```
if (isLoading && loadingElement !== undefined) {
    return loadingElement as React.JSX.Element; // because <StorageImage returns React.JSX.Element;
 }
```

```
interface StorageImageProps {
    ...,
    loadingElement?: React.ReactNode;
}
```

File: packages/react-storage/src/components/StorageImage/StorageImage.tsx (hunk 1 above)
Comment (dbanksdesign) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx]: I think the PR is good as is. The customer could pass `null` to `loadingElement` and it would effectively work the way you are describing @calebpollman ?

Comment (calebpollman) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx]: That's true @dbanksdesign. My concern here is that existing consumers will still have the issue that this PR is meant to address as there is no default `loadingElement`

File: docs/src/pages/[platform]/connected-components/storage/storage-browser/react.mdx [hunk 3]
```diff
@@ -27,6 +27,27 @@ Storage Browser for Amazon S3 is an open source component that you can add to yo
 
 End users work with S3 _locations_ within the Storage Browser interface. _Locations_ are S3 buckets or prefixes that you authorize end users to access using Amazon S3 Access Grants or Identity and Access Management (IAM) policies, depending on your use case. When the Storage Browser component is first rendered, it will show the LocationsView which displays only the locations you have granted them access to. Once an end user has selected a location, they can browse the S3 bucket or prefix, and all the data contained further down the S3 resource path, but they cannot browse buckets or prefixes higher up the S3 resource path.
 
+## Quick start
```
Comment (AllanZhengYP) [inline docs/src/pages/[platform]/connected-components/storage/storage-browser/react.mdx:30]: This diff should already be merged to main. Not sure why it eixsts here

File: examples/next-app-router/package.json [hunk 4]
```diff
@@ -11,7 +11,7 @@
   "dependencies": {
     "@aws-amplify/ui-react": "^6.1.0",
     "react": "^18.3.0",
-    "next": "^14.2.21",
+    "next": "^14.2.25",
```
Comment (AllanZhengYP) [inline examples/next-app-router/package.json:14]: This diff is from other PR as well. Can you rebase your PR?

Comment (aadimch) [reply examples/next-app-router/package.json:14]: Thank you, I will do them once [this conversation](https://github.com/aws-amplify/amplify-ui/pull/6448#discussion_r2019432838) is resolved

File: packages/react-storage/src/components/StorageImage/StorageImage.tsx [hunk 5]
```diff
@@ -99,7 +100,15 @@ export const StorageImage = ({
...
-  const { url } = useGetUrl(input);
+  const { url, isLoading } = useGetUrl(input);
+
+  if (isLoading && loadingElement) {
+    return <>{loadingElement}</>;
+  }
+
+  if (isLoading) {
+    return <>{null}</>;
+  }
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageImage/StorageImage.tsx:111]: These can be combined in single conditional, example:
```tsx
if (isLoading) {
  return !loadingElement ? null : <>{loadingElement}</>;
}
```

Also we probably need to update the return type of `StorageImage` to allow `null` as welll

Comment (aadimch) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx:111]: Couldn't updating the return type break something in the existing code, if it assumes the StorageImage to always return non-null value?

Comment (calebpollman) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx:111]: @aadimch Good question! In this case we are loosening the return type which is not a breaking change for a React component

Comment (aadimch) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx]: Makes sense, thank you. I have updated the PR, with the suggested changes.

File: packages/react-storage/src/components/StorageImage/StorageImage.tsx [hunk 6]
```diff
@@ -99,7 +100,11 @@ export const StorageImage = ({
     ]
   );
 
-  const { url } = useGetUrl(input);
+  const { url, isLoading } = useGetUrl(input);
+
+  if (isLoading) {
+    return !loadingElement ? null : <>{loadingElement}</>;
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageImage/StorageImage.tsx:106]: Can we add a unit test for the `null` return scenario?

Comment (aadimch) [reply packages/react-storage/src/components/StorageImage/StorageImage.tsx:106]: Added. Thank you.

**Review by calebpollman — APPROVED** [review]
🚢

**Review by dbanksdesign — DISMISSED** [review]

File: .changeset/thin-jars-count.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+"@aws-amplify/ui-docs": patch
```
Comment (calebpollman) [inline .changeset/thin-jars-count.md:2]: The docs don't need to be included since they aren't a published package
```suggestion
```

**Review by calebpollman — APPROVED** [review]

**Review by ashika112 — APPROVED** [review]

**Review by dbanksdesign — APPROVED** [review]
